### DISCUSSION TITLE
Changelog for loophp/collection 7

### DISCUSSION THREAD

**Discussion by drupol** [loophp/collection#284] [General]
Hello,

I'm opening the discussion for drafting the upcoming changelog for the upcoming version 7 of `loophp/collection` :tada: :partying_face: 

Estimate date of release: 11 December 2022

## Overall feeling
A year without update.

Released on the 11th December 2021, 6.0.3 was the last update.

Today, we are pleased to announce the release of version 7, our first update in over a year. There were several reasons for the delay, including a lack of bug reports and the busy schedules of our team members. We also wanted to take the time to carefully test and prepare the release, given the significant changes and improvements included in this update.

We hope you enjoy this update, and we look forward to continuing to improve and evolve our library in the coming months and years. Please let us know if you have any feedback or suggestions.

## Notable changes
- Minimum version of PHP is now 8
  PHP 7.4 support has ended the 28 November 2022, therefore, it's time to support PHP 8. [Rector][rectorphp repository] has been used to upgrade the codebase with all the new PHP 8 stuff.
- Removal of docker, switch to [Nix][nix website]
  We were using a docker-compose file to have the documentation built locally, but since I'm using more and more [Nix](https://nixos.org/), I decided to get rid of Docker. With one single file [`.envrc`](https://github.com/loophp/collection/blob/master/.envrc), all the dependencies we need to built the documentation server, creating changelogs with auto-changelogs, running Sphinx server for documentation and have a proper PHP interpreter.
- Removal of PHPSpec, switch to PHPunit
  For many reasons, we decided to switch to PHPUnit for testing. Switching to the PHPUnit framework was the opportunity to rewrite all the tests and we are extremelly grateful towards the PHPUnit framework. We also built a specific [PHPUnit plugin](https://github.com/loophp/phpunit-iterable-assertions) providing assertions for iterables. To built it, we had to abstract the main classes in use in `loophp/collection` in their [own project](https://github.com/loophp/iterators), and that project contains now many Iterator aggregates now in use in `loophp/collection`. 
- New `CollectionDecorator` abstract class to create custom collection classes
  Since the beginning of the creation of this library, many people expressed the will to not set the main Collection class as `final`. While I'm absolutely not for that, I'm more keen to provide an `abstract` class which acts as a decorator and that can be easily extended by using the inheritance mechanism.
- Add new operations: `Averages`, `Compare`, `Max`, `Min`
  A couple of new operations has been added, please read the API documentation to learn more about them.
- Use of `IteratorAggregate`
  Many operations are now using `IteratorAggregate` from `loophp/iterators`.
- The `cache` operation is using the fastest implementation of Caching iterators existing. I couldn't find a fastest implementation yet. That implementation is available in `loophp/iterators` as an Iterator aggregate. Special thanks to @azjezz for the discussion and tips.  
- Static analysis coverage has been greatly improved, we are now around 99.2%

## BC Breaking changes (`git log 6.0.3.. --grep="BREAKING"`)
- #269
  The static constructor is no more variadic. We noticed that when using `unfold`, there were some of useless kind of back and forth when it comes to callback arguments handling. Return value is an array, which is expanded, then converted again into an array... that mambo jambo is useless, it has been removed.  
- #245
  The `ScanLeft1` has always been something that was bugging me out in the edge cases. I was unable to find an elegant way to deal with it when the Collection was empty since we were using `Prepend` in it. This issue has been fixed now. The implementation is not the best, but it works very well.
- #249
  The operations:  `first`, `foldLeft`, `foldLeft1`, `foldRight`, `foldRight1`, `implode`, `reduce`, `unlines`, `unwords`, `last`, `head`, `get`, `compare`, `min`, `max` are now returning a value instead of returning a Collection containing one single value. That feature request has been requested many times ago by many users, this is now done.
- The `Tails` operation is now keeping keys.
  The `Tails` operation used to return a list of values, it is now returning value keys has well. The implementation is way cleaner and faster has well.

## Numbers
- 233 commits (`git log 6.0.3.. --pretty=oneline | wc -l`)
- 18 issues closed
- No user contribution #SadPandaFace
- ~80k downloads in total (~200 per day) (https://packagist.org/packages/loophp/collection/stats)
 
## Random
- [Advent Of Code 2022][aoc website]
  This year I decided to participate to the Advent Of Code and guess what...  I'm using `loophp/collection` for finding solutions to the challenges. It turns out that this is actually an excellent tool. I also have in mind to create a dedicated documentation section with the solutions when the challenge is over. In the meantime, you can have a look at the [aoc2022][aoc2022 repository] repository and get some inspirations.
- Try It Online!
  A new badge "Try It Online" on the README has been added. It uses https://play.phpsandbox.io/ and provides a live playground where you can test the library online in a full blown text editor with auto-completion !

[nix website]: https://nixos.org/
[rectorphp repository]: https://github.com/rectorphp/rector
[aoc2022 repository]: https://github.com/drupol/aoc2022
[aoc website]: https://adventofcode.com/

**Comment by AlexandruGG** [comment]
Hey @drupol, this looks good! I have two suggestions:

1. We should have a bullet point for the **BC-breaking changes**, including:
- https://github.com/loophp/collection/pull/269
- https://github.com/loophp/collection/pull/245
- https://github.com/loophp/collection/pull/259
- any others you can think of? (I couldn't find anything else)

2. With the drop of support for PHP 7.4, we can make sure the code is fully utilising PHP 8 syntax and automatically refactor with a tool like [Rector](https://github.com/rectorphp/rector). Unless you've already done this? If not, I'd be happy to look at this :)

**Reply by drupol** [reply]
Good feedback ! I will amend the message.

I indeed used Rector for the upgrade to PHP 8 :)

**Reply by drupol** [reply]
Changelog updated!

**Comment by jdreesen** [comment]
Very nice! But there's a small mistake: the date is wrong, also in the actual release notes.

"Released on the 11th December ~~2022~~ 2021, 6.0.3 was the last update."

**Reply by drupol** [reply]
Thanks mate, I fixed in both places.
